src/main/java/org/spongepowered/api/event/cause/Cause.java

```diff
@@ -467,111 +393,75 @@ public String toString() {
         String causeString = "Cause[";
         StringJoiner joiner = new StringJoiner(", ");
         for (int i = 0; i < this.cause.length; i++) {
-            joiner.add("{Name=" + this.names[i] + ", Object={" + this.cause[i].toString() + "}}");
+            joiner.add(this.cause[i].toString());
         }
         return causeString + joiner.toString() + "]";
     }
 
-    public static final class Builder implements ResettableBuilder<Cause, Builder> {
-
-        List<NamedCause> causes = new ArrayList<>();
-        Set<String> namesUsed = new HashSet<>();
+    private class Itr implements Iterator<Object> {
 
-        Builder() {
+        private int index = 0;
 
+        public Itr() {
         }
 
-        Builder source(Object object) {
-            this.causes.add(NamedCause.source(checkNotNull(object, "Source cannot be null!")));
-            this.namesUsed.add(NamedCause.SOURCE);
-            return this;
+        @Override
+        public Object next() {
+            if (this.index >= Cause.this.cause.length) {
+                throw new NoSuchElementException();
+            }
+            return Cause.this.cause[this.index++];
         }
 
-        public Builder owner(Object object) {
-            checkArgument(!this.namesUsed.contains(NamedCause.OWNER), "Already contains an owner!");
-            this.causes.add(NamedCause.owner(object));
-            this.namesUsed.add(NamedCause.OWNER);
-            return this;
+        @Override
+        public boolean hasNext() {
+            return this.index != Cause.this.cause.length;
         }
 
-        public Builder notifier(Object object) {
-            checkArgument(!this.namesUsed.contains(NamedCause.NOTIFIER), "Already contains a notifier!");
-            this.causes.add(NamedCause.notifier(object));
-            this.namesUsed.add(NamedCause.NOTIFIER);
-            return this;
-        }
+    }
+
+    public static final class Builder implements ResettableBuilder<Cause, Builder> {
+
+        List<Object> causes = new ArrayList<>();
```

Comment:
Its unlikely that it will resize very often so I'd prefer this over a linked structure I think.